pkg/solana/chainwriter/chain_writer.go

```diff
@@ -0,0 +1,452 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+
+	"github.com/gagliardetto/solana-go"
+	addresslookuptable "github.com/gagliardetto/solana-go/programs/address-lookup-table"
+	"github.com/gagliardetto/solana-go/rpc"
+
+	commoncodec "github.com/smartcontractkit/chainlink-common/pkg/codec"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/client"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/codec"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/fees"
+	"github.com/smartcontractkit/chainlink-solana/pkg/solana/txm"
+)
+
+const ServiceName = "SolanaChainWriter"
+
+type SolanaChainWriterService struct {
+	lggr   logger.Logger
+	reader client.Reader
+	txm    txm.TxManager
+	ge     fees.Estimator
+	config ChainWriterConfig
+
+	parsed  *codec.ParsedTypes
+	encoder types.Encoder
+
+	services.StateMachine
+}
+
+var (
+	_ services.Service     = &SolanaChainWriterService{}
+	_ types.ContractWriter = &SolanaChainWriterService{}
+)
+
+// nolint // ignoring naming suggestion
+type ChainWriterConfig struct {
+	Programs map[string]ProgramConfig
+}
+
+type ProgramConfig struct {
+	Methods map[string]MethodConfig
+	IDL     string
+}
+
+type MethodConfig struct {
+	FromAddress        string
+	InputModifications commoncodec.ModifiersConfig
+	ChainSpecificName  string
+	LookupTables       LookupTables
+	Accounts           []Lookup
+	// Location in the args where the debug ID is stored
+	DebugIDLocation string
+}
+
+func NewSolanaChainWriterService(logger logger.Logger, reader client.Reader, txm txm.TxManager, ge fees.Estimator, config ChainWriterConfig) (*SolanaChainWriterService, error) {
+	w := SolanaChainWriterService{
+		lggr:   logger,
+		reader: reader,
+		txm:    txm,
+		ge:     ge,
+		config: config,
+		parsed: &codec.ParsedTypes{EncoderDefs: map[string]codec.Entry{}, DecoderDefs: map[string]codec.Entry{}},
+	}
+
+	if err := w.parsePrograms(config); err != nil {
+		return nil, fmt.Errorf("failed to parse programs: %w", err)
+	}
+
+	var err error
+	if w.encoder, err = w.parsed.ToCodec(); err != nil {
+		return nil, fmt.Errorf("%w: failed to create codec", err)
+	}
+
+	return &w, nil
+}
+
+func (s *SolanaChainWriterService) parsePrograms(config ChainWriterConfig) error {
+	for program, programConfig := range config.Programs {
+		var idl codec.IDL
+		if err := json.Unmarshal([]byte(programConfig.IDL), &idl); err != nil {
+			return fmt.Errorf("failed to unmarshal IDL for program: %s, error: %w", program, err)
+		}
+		for method, methodConfig := range programConfig.Methods {
+			idlDef, err := codec.FindDefinitionFromIDL(codec.ChainConfigTypeInstructionDef, methodConfig.ChainSpecificName, idl)
+			if err != nil {
+				return err
+			}
+
+			inputMod, err := methodConfig.InputModifications.ToModifier(codec.DecoderHooks...)
```

Comment:
Noticed you made some changes to this method and we're not checking if `methodConfig.InputModifications` is nil or not anymore. Was there a change to the interface or validation that we don't need to do that here anymore?

Reply:
That's a good catch, I didn't notice it before because the tests run fine with `nil` `InputModifications`. I believe the reason for this is that a `nil` slice is Go can still have operations run on it, so `.ToModifier` is called without a nil pointer exception and in the context of the codec it is completely valid to have no modifiers